pkg/controller/deployment/util/deployment_util.go

```diff
@@ -56,6 +58,11 @@ const (
 	// is deployment.spec.replicas + maxSurge. Used by the underlying replica sets to estimate their
 	// proportions in case the deployment has surge replicas.
 	MaxReplicasAnnotation = "deployment.kubernetes.io/max-replicas"
+	// ReplicaSetReplicasBeforeScaleAnnotation is the number of replicas a replica set had before scaling began.
+	// The presence of this annotation indicates that a replica set has not yet fully scaled.
+	// The value is used by the underlying replica sets to estimate their scale proportions if the deployment has surge replicas
+	// (e.g. terminating pods) and the replica set is partially scaled due to the deployment's .spec.podReplacementPolicy.
+	ReplicaSetReplicasBeforeScaleAnnotation = "deployment.kubernetes.io/replicaset-replicas-before-scale"
```

Comment:
It's not clear to me why we need a new annotation on replicaset to decide how to distribute pods from a scale up across existing replicasets, especially with the skew issues that introduces to kubectl and the reconciliation issues for the annotation that introduces.

When the deployment in question has opted into TerminationComplete replacement policy, doesn't computing scaleBase as max(rs.Spec.Replicas, rs.Status.TerminatingReplicas) let us avoid overscaling without needing a new annotation?

Reply:
Also, even if we do need an annotation, shouldn't all the logic around setting / using / reconciling / clearing ReplicaSetReplicasBeforeScaleAnnotation only kick in when the deployment opted into TerminationComplete? The logic is really complicated, avoiding it unless absolutely required seems better.

Reply:
Deployments support proportional scaling and we need this annotation to preserve that behavior. The changes to the proportional scaling logic are outlined in the KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md#deployment-scaling-changes-and-a-new-annotation-for-replicasets

Since the scaling can be gradual, we have no way of inferring what `replicasBeforeScale` were before for all replicasets.

Using `rs.Status.TerminatingReplicas` does not work well in the scale computation:
- `TerminatingReplicas` are transient and can be removed by the replicaset controller anytime.
- We actually do not know how many terminating replicas each replicaset can have at each single point in time. The lower bound is 0, but there is no upper bound.

We have also looked at the kubectl skew impact and it is minimal: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md#kubectl-skew

Reply:
> Also, even if we do need an annotation, shouldn't all the logic around setting / using / reconciling / clearing ReplicaSetReplicasBeforeScaleAnnotation only kick in when the deployment opted into TerminationComplete? 

Similar to NewRSNewReplicas, this is done in the code above via the partialScaling variable. Should I duplicate the check as well in the `ComputeReplicaSetScaleAnnotations` function?

Clearing the annotation should be done when this feature is no longer used IMO.

Reply:
The code paths are difficult enough that I *really* don't think we should change them for all pods. Limiting new logic or new sync passes to pods which have opted into the new reconcile behavior would build confidence that a bug here would not impact existing pods / replicasets / deployments on upgrade.

(this is trying to learn lessons from issues with the sidecar container feature that impacted all pods, even ones that were not using the sidecards feature)

Reply:
It should be passed from above, but I agree it is safer to guard the logic in place and check if the feature is opted in. I will fix this. The clearing of the annotation should be unconditional, I think, to clean up after disabling or opting out of the feature.